File: libs/tax-mcp/src/lib/tools/expenses.ts
import { McpServer } from '@modelcontextprotocol/sdk/server/mcp.js';
import { z } from 'zod';
import { expenseEntriesCol, db } from '@can-tax-pro/db';
import { createExpenseEntrySchema, computeAmountCad } from '@can-tax-pro/utils';
import { ExpenseCategoryType, Currency } from '@can-tax-pro/types';
import { FieldValue } from 'firebase-admin/firestore';

export function registerExpenseTools(server: McpServer, userId: string) {
  server.tool(
    'list_expenses',
    'List business expense entries for a tax year, optionally filtered by category',
    {
      taxYearId: z.string().describe('Firestore tax year document ID'),
      category: z.nativeEnum(ExpenseCategoryType).optional(),
    },
    async ({ taxYearId, category }) => {
      let q = expenseEntriesCol(userId, taxYearId).orderBy('date', 'desc') as FirebaseFirestore.Query;
      if (category) q = q.where('category', '==', category);
      const snap = await q.get();
      const entries = snap.docs.map((d) => ({ id: d.id, ...d.data() }));
      return { content: [{ type: 'text', text: JSON.stringify(entries, null, 2) }] };
    }
  );

  server.tool(
    'create_expense',
    'Create a single business expense entry for a tax year',
    {
      taxYearId: z.string(),
      category: z.nativeEnum(ExpenseCategoryType),
      vendor: z.string().optional(),
      description: z.string().optional(),
      amount: z.number().positive(),
      currency: z.nativeEnum(Currency).optional(),
      exchangeRate: z.number().positive().optional(),
      date: z.string().date().describe('ISO date string YYYY-MM-DD'),
      paymentMethod: z.string().optional(),
      metadata: z.record(z.string(), z.unknown()).optional(),
    },
    async ({ taxYearId, ...data }) => {
      const parsed = createExpenseEntrySchema.safeParse(data);
      if (!parsed.success) throw new Error(parsed.error.message);
      const currency = parsed.data.currency ?? Currency.CAD;
      const amountCad = computeAmountCad(parsed.data.amount, currency, parsed.data.exchangeRate);
      const ref = await expenseEntriesCol(userId, taxYearId).add({
        ...parsed.data,
        currency,
        amountCad,
        date: new Date(parsed.data.date),
        createdAt: FieldValue.serverTimestamp(),
        updatedAt: FieldValue.serverTimestamp(),
      });
      const doc = await ref.get();
      return { content: [{ type: 'text', text: JSON.stringify({ id: doc.id, ...doc.data() }) }] };
    }
  );

  server.tool(
    'bulk_import_expenses',
    'Batch-import multiple business expense entries at once for a tax year',
    {
      taxYearId: z.string(),
      entries: z.array(createExpenseEntrySchema).min(1),
    },
    async ({ taxYearId, entries }) => {
      const batch = db.batch();
      const col = expenseEntriesCol(userId, taxYearId);
      for (const entry of entries) {
        const currency = entry.currency ?? Currency.CAD;
        const amountCad = computeAmountCad(entry.amount, currency, entry.exchangeRate);
        batch.set(col.doc(), {
          ...entry,
          currency,
          amountCad,
          date: new Date(entry.date),
          createdAt: FieldValue.serverTimestamp(),
          updatedAt: FieldValue.serverTimestamp(),
        });
      }
      await batch.commit();
      return { content: [{ type: 'text', text: JSON.stringify({ imported: entries.length }) }] };
    }
  );
}

File: tax-mcp/tools/expenses.py
import json
from datetime import datetime, timezone
from typing import Optional
from google.cloud.firestore import Client
from mcp.server.fastmcp import FastMCP


def _compute_amount_cad(amount: float, currency: str, exchange_rate: Optional[float]) -> float:
    if currency == "CAD":
        return amount
    if not exchange_rate:
        raise ValueError("exchangeRate required for non-CAD currency")
    return round(amount * exchange_rate, 2)


def _expense_col(db: Client, user_id: str, tax_year_id: str):
    return (
        db.collection("users").document(user_id)
        .collection("taxYears").document(tax_year_id)
        .collection("expenseEntries")
    )


def register_expense_tools(mcp: FastMCP, db: Client, user_id: str):

    @mcp.tool()
    def list_expenses(tax_year_id: str, category: Optional[str] = None) -> str:
        """List expense entries. category: EMAIL | GCP | NAMECHEAP | PHONE | INTERNET | ADS | HOSTING | SOFTWARE | OTHER"""
        q = _expense_col(db, user_id, tax_year_id).order_by("date", direction="DESCENDING")
        if category:
            q = q.where("category", "==", category)
        return json.dumps([{"id": d.id, **d.to_dict()} for d in q.stream()], default=str)

    @mcp.tool()
    def create_expense(
        tax_year_id: str,
        category: str,
        amount: float,
        date: str,
        currency: str = "CAD",
        exchange_rate: Optional[float] = None,
        vendor: Optional[str] = None,
        description: Optional[str] = None,
        payment_method: Optional[str] = None,
    ) -> str:
        """Create a single expense entry. date: YYYY-MM-DD. category: EMAIL | GCP | NAMECHEAP | PHONE | INTERNET | ADS | HOSTING | SOFTWARE | OTHER"""
        amount_cad = _compute_amount_cad(amount, currency, exchange_rate)
        now = datetime.now(timezone.utc)
        data = {
            "category": category,
            "amount": amount,
            "currency": currency,
            "exchangeRate": exchange_rate,
            "amountCad": amount_cad,
            "date": datetime.fromisoformat(date),
            "vendor": vendor,
            "description": description,
            "paymentMethod": payment_method,
            "createdAt": now,
            "updatedAt": now,
        }
        ref = _expense_col(db, user_id, tax_year_id).document()
        ref.set({k: v for k, v in data.items() if v is not None})
        doc = ref.get()
        return json.dumps({"id": doc.id, **doc.to_dict()}, default=str)

    @mcp.tool()
    def delete_expense(tax_year_id: str, expense_id: str) -> str:
        """Delete a single expense entry by ID."""
        _expense_col(db, user_id, tax_year_id).document(expense_id).delete()
        return json.dumps({"deleted": expense_id})

    @mcp.tool()
    def bulk_import_expenses(tax_year_id: str, entries: str) -> str:
        """Batch-import multiple expense entries. entries: JSON string array, each item: {category, amount, date, currency?, exchangeRate?, vendor?, description?}"""
        parsed: list[dict] = json.loads(entries)
        col = _expense_col(db, user_id, tax_year_id)
        batch = db.batch()
        now = datetime.now(timezone.utc)
        for entry in parsed:
            currency = entry.get("currency", "CAD")
            amount_cad = _compute_amount_cad(entry["amount"], currency, entry.get("exchangeRate"))
            ref = col.document()
            batch.set(ref, {
                "category": entry["category"],
                "amount": entry["amount"],
                "currency": currency,
                "exchangeRate": entry.get("exchangeRate"),
                "amountCad": amount_cad,
                "date": datetime.fromisoformat(entry["date"]),
                "vendor": entry.get("vendor"),
                "description": entry.get("description"),
                "paymentMethod": entry.get("paymentMethod"),
                "createdAt": now,
                "updatedAt": now,
            })
        batch.commit()
        return json.dumps({"imported": len(parsed)})
